anvil/tests/test_models.py

```diff
@@ -53,3 +55,45 @@ def test_model_construction(layer_idx, n_blocks, lattice_length_half, hidden_sha
     # might help with memory.
     with torch.no_grad():
         API.model_to_load(**params)
+
+
+def layer_independence_test(model_spec):
+    """Check that each layer's parameters are updated independently."""
+
+    # Collect over these layers
+    model = API.model_to_load(**model_spec)
+    layer1, layer2 = [layer for layer in model]
+
+    layer2_copy = deepcopy(layer2)
+
+    # Update parameters in first layer
+    valid_key, valid_tensor = next(iter(layer1.state_dict().items()))
+    update = {valid_key: torch.rand_like(valid_tensor)}
+    layer1.load_state_dict(update, strict=False)
+
+    # Check that second layer is unchanged
+    # NOTE: may be safer to iterate over shared keys
+    for original, copy in zip(layer2.parameters(), layer2_copy.parameters()):
+        assert torch.allclose(original, copy)
+
+
+# TODO: extend to other layers... @pytest.mark.parametrize("layer_action", LAYERS)
+@torch.no_grad()
+def test_layer_independence_global_rescaling():
+    # Build a model with two identical sets of layers
+    working_example = {  # This is OK
+        "model": [
+            {"layer": "global_rescaling", "scale": 1.0},
+            {"layer": "global_rescaling", "scale": 1.0},
+        ]
+    }
+    layer_independence_test(working_example)
+
+    breaking_example = {  # This is NOT ok!
+        "model": [
+            {"layer": "global_rescaling"},
+            {"layer": "global_rescaling"},
+        ],
+        "scale": 1.0,
+    }
```

Comment:
This example fails the test, i.e. when you update layer 1, layer 2's parameters are also updated. I'm not sure if this is expected behaviour from reportengine or not, but it's the same story if we just fall back on default values.

Reply:
if you want a breaking example then use https://docs.pytest.org/en/6.2.x/assert.html#assertions-about-expected-exceptions

Reply:
Also don't forget you can use the API to check that the thing reportengine is doing is correct (it let's you get object from the resource builder and therefore ensures that higher level behaviour is correct)

Reply:
I wonder if this gets fixed by https://github.com/NNPDF/reportengine/pull/47 btw. Before that RE basically fails to re-evaluate certain things when doing collect because of over assuming how constant the results of productions rules are. But with that new version it forces it to properly resolve things every time..